File: examples/dekita/.claude/hooks/issue-comments-check.py
#!/usr/bin/env python3
"""gh issue viewコマンド実行時にコメントを自動表示する。

Why:
    Issueコメントに重要な解決策や追加情報があっても見落とされ、
    無駄な時間を費やすことがある。コメントを自動表示して
    情報の見落としを防ぐ。

What:
    - gh issue view <number> コマンドを検出
    - --commentsフラグがない場合、自動でコメントを取得
    - systemMessageでコメント内容を表示

Remarks:
    - 非ブロック型（情報提供のみ）
    - --comments付きのコマンドはそのまま通過

Changelog:
    - silenvx/dekita#538: フック追加（コメント見落とし防止）
"""

import json
import re
import subprocess

from lib.constants import TIMEOUT_MEDIUM
from lib.execution import log_hook_execution
from lib.results import print_continue_and_log_skip
from lib.session import create_hook_context, parse_hook_input
from lib.strings import strip_quoted_strings


def extract_issue_number(command: str) -> str | None:
    """Extract issue number from gh issue view command.

    Args:
        command: The bash command string.

    Returns:
        Issue number as string, or None if not found.

    Handles various flag positions:
    - gh issue view 123
    - gh issue view #123
    - gh issue view --web 123
    - gh issue view 123 --web
    """
    # Remove quoted strings to avoid false positives
    cmd = strip_quoted_strings(command)

    # Check if this is a gh issue view command
    if not re.search(r"gh\s+issue\s+view\b", cmd):
        return None

    # Extract all arguments after "gh issue view"
    match = re.search(r"gh\s+issue\s+view\s+(.+)", cmd)
    if not match:
        return None

    args = match.group(1)

    # Find issue number (with or without #) among the arguments
    # Skip flags (--flag or -f) and their values
    for part in args.split():
        # Skip flags and flag values
        if part.startswith("-"):
            continue
        # Match issue number (with optional # prefix)
        num_match = re.match(r"#?(\d+)$", part)
        if num_match:
            return num_match.group(1)

    return None


def has_comments_flag(command: str) -> bool:
    """Check if command already has --comments flag.

    Args:
        command: The bash command string.

    Returns:
        True if --comments flag is present as a standalone flag.
    """
    # Remove quoted strings to avoid matching flags inside quotes
    cmd = strip_quoted_strings(command)
    # Match --comments as a standalone flag (bounded by start/end or whitespace)
    return re.search(r"(?:^|\s)--comments(?:\s|$)", cmd) is not None


def fetch_issue_comments(issue_number: str) -> tuple[bool, str]:
    """Fetch issue comments using gh CLI.

    Args:
        issue_number: The issue number.

    Returns:
        Tuple of (success, comments):
        - (True, comments) if successful with comments
        - (True, "") if successful but no comments
        - (False, "") if error occurred
    """
    try:
        result = subprocess.run(
            [
                "gh",
                "issue",
                "view",
                issue_number,
                "--json",
                "comments",
                "--jq",
                '.comments[] | "---\\n**" + .author.login + "** (" + .createdAt[:10] + "):\\n" + .body + "\\n"',
            ],
            capture_output=True,
            text=True,
            timeout=TIMEOUT_MEDIUM,
            check=False,
        )

        if result.returncode != 0:
            return (False, "")

        return (True, result.stdout.strip())

    except (subprocess.TimeoutExpired, OSError):
        return (False, "")


def main():
    """
    PreToolUse hook for Bash commands.

    Automatically fetches and displays issue comments when `gh issue view`
    is called without --comments flag.
    """
    result = {"decision": "approve"}

    try:
        input_data = parse_hook_input()
        # Issue #2607: Create context for session_id logging
        ctx = create_hook_context(input_data)
        tool_name = input_data.get("tool_name", "")

        if tool_name != "Bash":
            print_continue_and_log_skip("issue-comments-check", f"not Bash: {tool_name}", ctx=ctx)
            return

        tool_input = input_data.get("tool_input", {})
        command = tool_input.get("command", "")

        # Check if this is a gh issue view command
        issue_number = extract_issue_number(command)
        if not issue_number:
            print_continue_and_log_skip("issue-comments-check", "no issue number found", ctx=ctx)
            return

        # If --comments is already present, let it through
        if has_comments_flag(command):
            log_hook_execution(
                "issue-comments-check",
                "approve",
                f"--comments付き: Issue #{issue_number}",
            )
            print(json.dumps(result))
            return

        # Fetch comments and display via systemMessage
        success, comments = fetch_issue_comments(issue_number)

        if not success:
            # Don't show misleading message on error
            log_hook_execution(
                "issue-comments-check",
                "approve",
                f"コメント取得エラー: Issue #{issue_number}",
            )
        elif comments:
            result["systemMessage"] = (
                f"📝 **Issue #{issue_number} のコメント** (自動取得)\n\n"
                f"{comments}\n\n"
                f"💡 Issueに取り組む前に、必ずコメントを確認してください。"
            )
            log_hook_execution(
                "issue-comments-check",
                "approve",
                f"コメント自動表示: Issue #{issue_number}",
            )
        else:
            result["systemMessage"] = f"ℹ️ Issue #{issue_number} にはコメントがありません。"
            log_hook_execution(
                "issue-comments-check",
                "approve",
                f"コメントなし: Issue #{issue_number}",
            )

    except Exception as e:
        # Don't block on errors
        log_hook_execution(
            "issue-comments-check",
            "error",
            f"フックエラー: {e}",
        )

    print(json.dumps(result))


if __name__ == "__main__":
    main()

File: examples/dekita/.claude/hooks/handlers/issue_comments_check.ts
#!/usr/bin/env bun
/**
 * gh issue viewコマンド実行時にコメントを自動表示する。
 *
 * Why:
 *   Issueコメントに重要な解決策や追加情報があっても見落とされ、
 *   無駄な時間を費やすことがある。コメントを自動表示して
 *   情報の見落としを防ぐ。
 *
 * What:
 *   - gh issue view <number> コマンドを検出
 *   - --commentsフラグがない場合、自動でコメントを取得
 *   - systemMessageでコメント内容を表示
 *
 * Remarks:
 *   - 非ブロック型（情報提供のみ）
 *   - --comments付きのコマンドはそのまま通過
 *   - Python版: issue_comments_check.py
 *
 * Changelog:
 *   - silenvx/dekita#538: フック追加（コメント見落とし防止）
 *   - silenvx/dekita#2917: TypeScript版初期実装
 */

import { execSync } from "node:child_process";
import { TIMEOUT_MEDIUM } from "../lib/constants";
import { formatError } from "../lib/format_error";
import { logHookExecution } from "../lib/logging";
import { parseHookInput } from "../lib/session";
import { stripQuotedStrings } from "../lib/strings";

const HOOK_NAME = "issue-comments-check";

/**
 * Extract issue number from gh issue view command.
 *
 * Handles various flag positions:
 * - gh issue view 123
 * - gh issue view #123
 * - gh issue view --web 123
 * - gh issue view 123 --web
 */
export function extractIssueNumber(command: string): string | null {
  // Remove quoted strings to avoid false positives
  const cmd = stripQuotedStrings(command);

  // Check if this is a gh issue view command
  if (!/gh\s+issue\s+view\b/.test(cmd)) {
    return null;
  }

  // Extract all arguments after "gh issue view"
  const match = cmd.match(/gh\s+issue\s+view\s+(.+)/);
  if (!match) {
    return null;
  }

  const args = match[1];

  // Find issue number (with or without #) among the arguments
  // Skip flags (--flag or -f) and their values
  for (const part of args.split(/\s+/)) {
    // Skip flags and flag values
    if (part.startsWith("-")) {
      continue;
    }
    // Match issue number (with optional # prefix)
    const numMatch = part.match(/^#?(\d+)$/);
    if (numMatch) {
      return numMatch[1];
    }
  }

  return null;
}

/**
 * Check if command already has --comments flag.
 */
export function hasCommentsFlag(command: string): boolean {
  // Remove quoted strings to avoid matching flags inside quotes
  const cmd = stripQuotedStrings(command);
  // Match --comments as a standalone flag (bounded by start/end or whitespace)
  return /(?:^|\s)--comments(?:\s|$)/.test(cmd);
}

/**
 * Fetch issue comments using gh CLI.
 *
 * @returns [success, comments]:
 *   - [true, comments] if successful with comments
 *   - [true, ""] if successful but no comments
 *   - [false, ""] if error occurred
 */
function fetchIssueComments(issueNumber: string): [boolean, string] {
  try {
    const result = execSync(
      `gh issue view ${issueNumber} --json comments --jq '.comments[] | "---\\n**" + .author.login + "** (" + .createdAt[:10] + "):\\n" + .body + "\\n"'`,
      {
        encoding: "utf-8",
        timeout: TIMEOUT_MEDIUM * 1000,
        stdio: ["pipe", "pipe", "pipe"],
      },
    );
    return [true, result.trim()];
  } catch {
    // gh CLI not installed, timeout, or other error
    return [false, ""];
  }
}

interface ApproveResult {
  systemMessage?: string;
}

async function main(): Promise<void> {
  const result: ApproveResult = {};
  let sessionId: string | undefined;

  try {
    const data = await parseHookInput();
    sessionId = data.session_id;
    const toolName = (data.tool_name as string) || "";

    if (toolName !== "Bash") {
      await logHookExecution(HOOK_NAME, "approve", `not Bash: ${toolName}`, undefined, {
        sessionId,
      });
      console.log(JSON.stringify(result));
      return;
    }

    const toolInput = (data.tool_input as Record<string, unknown>) || {};
    const command = (toolInput.command as string) || "";

    // Check if this is a gh issue view command
    const issueNumber = extractIssueNumber(command);
    if (!issueNumber) {
      await logHookExecution(HOOK_NAME, "approve", "no issue number found", undefined, {
        sessionId,
      });
      console.log(JSON.stringify(result));
      return;
    }

    // If --comments is already present, let it through
    if (hasCommentsFlag(command)) {
      await logHookExecution(
        HOOK_NAME,
        "approve",
        `--comments付き: Issue #${issueNumber}`,
        undefined,
        { sessionId },
      );
      console.log(JSON.stringify(result));
      return;
    }

    // Fetch comments and display via systemMessage
    const [success, comments] = fetchIssueComments(issueNumber);

    if (!success) {
      // Don't show misleading message on error
      await logHookExecution(
        HOOK_NAME,
        "approve",
        `コメント取得エラー: Issue #${issueNumber}`,
        undefined,
        { sessionId },
      );
    } else if (comments) {
      result.systemMessage = `📝 **Issue #${issueNumber} のコメント** (自動取得)\n\n${comments}\n\n💡 Issueに取り組む前に、必ずコメントを確認してください。`;
      await logHookExecution(
        HOOK_NAME,
        "approve",
        `コメント自動表示: Issue #${issueNumber}`,
        undefined,
        { sessionId },
      );
    } else {
      result.systemMessage = `ℹ️ Issue #${issueNumber} にはコメントがありません。`;
      await logHookExecution(
        HOOK_NAME,
        "approve",
        `コメントなし: Issue #${issueNumber}`,
        undefined,
        { sessionId },
      );
    }
  } catch (error) {
    // Don't block on errors
    await logHookExecution(HOOK_NAME, "error", `フックエラー: ${formatError(error)}`, undefined, {
      sessionId,
    });
  }

  console.log(JSON.stringify(result));
}

if (import.meta.main) {
  main();
}
